src/node/timer.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/logger.h"
+#include "ds/spinlock.h"
+
+#include <chrono>
+#include <functional>
+#include <list>
+
+namespace ccf
+{
+  using TimerCallback = std::function<bool()>;
+
+  /**
+   * A timer class to trigger actions periodically.
+   *
+   * A shared_ptr to a new Timer is returned by Timers::new_timer(period, cb).
+   * The timer is only active when it is started (start()) and marked as expired
+   * when the callback is triggered. If the callback returns true, the timer
+   * continues ticking. Otherwise, the timer expires and will tick again only
+   * when it is explicitely re-started (start()).
+   *
+   **/
+  class Timer
+  {
+  private:
+    enum TimerState
+    {
+      STOPPED = 0,
+      STARTED,
+      EXPIRED
+    };
+
+    SpinLock lock;
+    std::chrono::milliseconds period;
+    std::chrono::milliseconds tick_;
+    TimerCallback cb;
+    TimerState state;
+
+  public:
+    Timer(std::chrono::milliseconds period_, TimerCallback cb_) :
+      period(period_),
+      tick_(0),
+      cb(cb_),
+      state(TimerState::STOPPED)
+    {}
+
+    void start()
+    {
+      std::lock_guard<SpinLock> guard(lock);
+      state = TimerState::STARTED;
+    }
+
+    void tick(std::chrono::milliseconds elapsed)
+    {
+      std::lock_guard<SpinLock> guard(lock);
+
+      if (state != TimerState::STARTED)
+        return;
+
+      tick_ += elapsed;
+      if (tick_ >= period)
+      {
+        state = TimerState::EXPIRED;
+        if (cb())
+          state = TimerState::STARTED;
+
+        using namespace std::chrono_literals;
+        tick_ = 0ms;
+      }
+    }
+  };
+
+  class Timers
+  {
+  private:
+    SpinLock lock;
+    std::list<std::shared_ptr<Timer>> timers;
```

Comment:
This design is fine for current purposes, but if we use this more widely I think we should change it slightly:
- `Timers` should store `weak_ptr`s, ticking those which are still owned elsewhere and removing any which become unreferenced
- `Timers::timers` should be a set
- The handle returned by `new_timer` should only expose `start`, and maybe `pause` - `tick` should only be visible to `Timers`
- Documentation should describe precise behaviour of `tick`. We expect it is called regularly, in small increments. Callbacks are currently only ever called once per `tick` (ie - if the period is `1ms`, and we get a tick of `10ms`, we don't restart the timer and instantly re-trigger it 10 times)

Reply:
This is now addressed.